drivers/led/ncp5623c.c

```diff
@@ -0,0 +1,213 @@
+
+#include "led_context.h"
+
+#define NCP5623_LED_CURRENT 0x20
+#define NCP5623_LED_PWM0    0x40
+#define NCP5623_LED_PWM1    0x60
+#define NCP5623_LED_PWM2    0x80
+
+#define NCP5623C_CHANNEL_COUNT 3
+
+/* Brightness limits */
+#define NCP5623C_MIN_BRIGHTNESS 0
+#define NCP5623C_MAX_BRIGHTNESS 0x1f
+
+static const uint8_t led_channels[] = {NCP5623_LED_PWM0, NCP5623_LED_PWM1, NCP5623_LED_PWM2};
+
+struct ncp5623c_config {
+	struct i2c_dt_spec bus;
+	uint8_t num_leds;
+	const struct led_info *leds_info;
+};
+
+static const struct led_info *ncp5623c_led_to_info(const struct ncp5623c_config *config,
+						   uint32_t led)
+{
+	if (led < config->num_leds) {
+		return &config->leds_info[led];
+	}
+
+	return NULL;
+}
+
+static int ncp5623c_get_info(const struct device *dev, uint32_t led, const struct led_info **info)
+{
+	const struct ncp5623c_config *config = dev->config;
+	const struct led_info *led_info = ncp5623c_led_to_info(config, led);
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	*info = led_info;
+
+	return 0;
+}
+
+static int ncp5623c_set_color(const struct device *dev, uint32_t led, uint8_t num_colors,
+			      const uint8_t *color)
+{
+	const struct ncp5623c_config *config = dev->config;
+	const struct led_info *led_info = ncp5623c_led_to_info(config, led);
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (led_info->num_colors != 3) {
+		return -ENOTSUP;
+	}
+	if (num_colors != 3) {
+		return -EINVAL;
+	}
+
+	uint8_t buf[6] = {0x70, NCP5623_LED_PWM0, 0x70, NCP5623_LED_PWM1, 0x70, NCP5623_LED_PWM2};
+
+	buf[1] = buf[1] | color[0] / 8;
+	buf[3] = buf[3] | color[1] / 8;
+	buf[5] = buf[5] | color[2] / 8;
+
+	return i2c_burst_write_dt(&config->bus, NCP5623_LED_CURRENT | NCP5623C_MAX_BRIGHTNESS, buf,
+				  sizeof(buf));
+}
+
+static int ncp5623c_set_brightness(const struct device *dev, uint32_t led, uint8_t value)
+{
+	const struct ncp5623c_config *config = dev->config;
+	const struct led_info *led_info = ncp5623c_led_to_info(config, led);
+	int ret = 0;
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (value > 100) {
+		return -EINVAL;
+	}
+
+	if (led_info->num_colors != 1) {
+		return -ENOTSUP;
+	}
+
+	/* Rescale 0..100 to 0..31 */
+	value = value * NCP5623C_MAX_BRIGHTNESS / 100;
+
+	ret = i2c_reg_write_byte_dt(&config->bus, led_channels[led] | value, 0x70);
+
+	if (ret < 0) {
+		LOG_ERR("LED write failed");
+	}
+
+	return ret;
+}
+
+static inline int ncp5623c_led_on(const struct device *dev, uint32_t led)
+{
+	return ncp5623c_set_brightness(dev, led, 100);
+}
+
+static inline int ncp5623c_led_off(const struct device *dev, uint32_t led)
+{
+	return ncp5623c_set_brightness(dev, led, 0);
+}
+
+static int ncp5623c_led_init(const struct device *dev)
+{
+	const struct ncp5623c_config *config = dev->config;
+	int i, num_color;
+
+	if (!i2c_is_ready_dt(&config->bus)) {
+		LOG_ERR("I2C device not ready");
+		return -ENODEV;
+	}
+
+	if (config->num_leds == 1) {
+		num_color = ncp5623c_led_to_info(config, 0)->num_colors;
```

Comment:
I know that the pointer returned by `ncp5623c_led_to_info()` can't be null but I don't like it. One day it might if the code around changes. Let's be more protective here.